plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs

```diff
@@ -323,19 +355,27 @@ instance Default ModelThreeArguments where
 
 runThreeArgumentModel
     :: ModelThreeArguments
-    -> ExMemory
-    -> ExMemory
-    -> ExMemory
-    -> CostingInteger
-runThreeArgumentModel (ModelThreeArgumentsConstantCost c) = lazy $ \_ _ _ -> c
-runThreeArgumentModel (ModelThreeArgumentsAddedSizes (ModelAddedSizes intercept slope)) = \(ExMemory size1) (ExMemory size2) (ExMemory size3) ->
-    (size1 + size2 + size3) * slope + intercept
-runThreeArgumentModel (ModelThreeArgumentsLinearInX (ModelLinearSize intercept slope)) = \(ExMemory size1) _ _ ->
-    size1 * slope + intercept
-runThreeArgumentModel (ModelThreeArgumentsLinearInY (ModelLinearSize intercept slope)) = \_ (ExMemory size2) _ ->
-    size2 * slope + intercept
-runThreeArgumentModel (ModelThreeArgumentsLinearInZ (ModelLinearSize intercept slope)) = \_ _ (ExMemory size3) ->
-    size3 * slope + intercept
+    -> CostStream
+    -> CostStream
+    -> CostStream
+    -> CostStream
+runThreeArgumentModel (ModelThreeArgumentsConstantCost c) = lazy $ \_ _ _ -> CostLast c
+runThreeArgumentModel
+    (ModelThreeArgumentsAddedSizes (ModelAddedSizes intercept slope)) =
+        lazy $ \costs1 costs2 costs3 ->
+            scaleLinearly intercept slope . addCostStream costs1 $ addCostStream costs2 costs3
```

Comment:
This is perhaps not as fast as if we had `addCostStream3` or some n-ary version with type-level tracking of the number of streams for unfolding purposes. Not sure if we care, though.